Lab-2/lab2_matrixfactorization.ipynb

```diff
@@ -0,0 +1,566 @@
+        "\n",
+        "def qrf(A):\n",
+        "\n",
+        "  assert(np.linalg.det(A) != 0)\n",
+        "\n",
+        "  Q = np.zeros(A.shape)\n",
+        "  R = np.zeros(A.shape)\n",
+        "\n",
+        "  V = A[:,[0]]\n",
+        "\n",
+        "  for i in range(A.shape[0]):\n",
+        "    R[i, i] = np.linalg.norm(V)\n",
+        "    Q[:, [i]] = V / R[i, i]\n",
+        "    for j in range(i+1, A.shape[0]):\n",
+        "      R[i, j] = np.sum(np.dot(Q[:,[i]].T, A[:,[j]]))\n",
+        "      V = A[:,[j]] - R[i, j]*Q[:,[i]]\n",
+        "\n",
+        "  return [Q, R]"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "fPft7QXJvvST",
+        "colab_type": "text"
+      },
+      "source": [
+        "We formulate a quick test to make sure that QR Factorization is working properly, an important property is that R is triangular:"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "o9pt0xjBsIzW",
+        "colab_type": "code",
+        "colab": {
+          "base_uri": "https://localhost:8080/",
+          "height": 153
+        },
+        "outputId": "bd331f37-094b-4870-a3ea-76b264f53309"
+      },
+      "source": [
+        "def is_triangular(A):\n",
+        "  for i in range(A.shape[0]):\n",
+        "    for j in range(0, i):\n",
+        "      if (A[i, j] != 0): return False\n",
+        "  return True\n",
+        "\n",
+        "def qrf_test():\n",
+        "  A = np.matrix(\"2 -1; -1 2\")\n",
+        "  Q, R = qrf(A)\n",
+        "  print(\"A = \", A)\n",
+        "  print(\"Q = \", Q)\n",
+        "  print(\"R = \", R)\n",
+        "  print(\"is_triangular(R) = \", is_triangular(R))\n",
+        "\n",
+        "qrf_test()\n"
+      ],
+      "execution_count": 102,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "text": [
+            "A =  [[ 2 -1]\n",
+            " [-1  2]]\n",
+            "Q =  [[ 0.89442719  0.4472136 ]\n",
+            " [-0.4472136   0.89442719]]\n",
+            "R =  [[ 2.23606798 -1.78885438]\n",
+            " [ 0.          1.34164079]]\n",
+            "is_triangular(R) =  True\n"
+          ],
+          "name": "stdout"
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "colab_type": "text",
+        "id": "lVE-mL2SDDZu"
+      },
+      "source": [
+        "### Direct Solver ###"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "I64B1TRbwBJQ",
+        "colab_type": "text"
+      },
+      "source": [
+        "From litterature we know that if Ax=b then x=A⁻1b, this is what is refered to as a direct solver technique:"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "colab_type": "code",
+        "id": "BBZekOW0DGxN",
+        "colab": {}
+      },
+      "source": [
+        "def ds(A, b):\n",
```

Comment:
You are here using numpy directly, which is not allowed. This assignment (assignment 3) needs to utilize the algorithms mentioned in the theory, see lecture notes part 3 for details.